### PR TITLE
Removes constraint in `BlockNumberProvider` from treasury

### PR DESCRIPTION
**Pull request by gupnik**
https://github.com/paritytech/polkadot-sdk/pull/3970 updated the treasury pallet to support relay chain block number provider. However, it added a constraint to the BlockNumberProvider to have the same block number type as frame_system:

```rust
type BlockNumberProvider: BlockNumberProvider<BlockNumber = BlockNumberFor<Self>>;
```

This PR removes that constraint as suggested by @gui1117

### CODE REVIEW

**Review by gui1117 — APPROVED** [review]

File: substrate/frame/treasury/src/lib.rs [hunk 1]
```diff
@@ -106,7 +106,7 @@ use frame_support::{
 	weights::Weight,
 	BoundedVec, PalletId,
 };
-use frame_system::pallet_prelude::BlockNumberFor;
+use frame_system::pallet_prelude::BlockNumberFor as SystemBlockNumberFor;
```
Comment (kianenigma) [inline substrate/frame/treasury/src/lib.rs:109]: Related to #6245, someday this type alias can be removed to avoid confusion.

**Review by kianenigma — APPROVED** [review]

File: substrate/frame/treasury/src/lib.rs [hunk 2]
```diff
@@ -122,6 +122,8 @@ pub type NegativeImbalanceOf<T, I = ()> = <<T as Config<I>>::Currency as Currenc
 >>::NegativeImbalance;
 type AccountIdLookupOf<T> = <<T as frame_system::Config>::Lookup as StaticLookup>::Source;
 type BeneficiaryLookupOf<T, I> = <<T as Config<I>>::BeneficiaryLookup as StaticLookup>::Source;
+pub type BlockNumberFor<T, I = ()> =
```
Comment (muharem) [inline substrate/frame/treasury/src/lib.rs:125]: not sure if we should use same naming. we have `BlockNumberFor` reference type in prelude that we use to reference the number from the header.

File: substrate/frame/treasury/src/lib.rs (hunk 1 above)
Comment (muharem) [reply substrate/frame/treasury/src/lib.rs:109]: May be local block number? What system really means here. Also `BlockNumberFor` reference to the block's header, not system pallet

File: substrate/frame/treasury/src/lib.rs (hunk 2 above)
Comment (muharem) [inline substrate/frame/treasury/src/lib.rs:125]: I would rather come up with new name, and not reuse the one which is used for local block number type

File: substrate/frame/treasury/src/lib.rs [hunk 3]
```diff
@@ -458,23 +460,23 @@ pub mod pallet {
...
+			if spend_periods_passed > BlockNumberFor::<T, I>::zero() {
 				Self::spend_funds(spend_periods_passed, new_last_spend_period)
 			} else {
 				Weight::zero()
 			}
 		}
 
 		#[cfg(feature = "try-runtime")]
-		fn try_state(_: BlockNumberFor<T>) -> Result<(), sp_runtime::TryRuntimeError> {
+		fn try_state(_: BlockNumberFor<T, I>) -> Result<(), sp_runtime::TryRuntimeError> {
```
Comment (muharem) [inline substrate/frame/treasury/src/lib.rs:479]: this should be local

File: substrate/frame/treasury/src/lib.rs (hunk 2 above)
Comment (gupnik) [reply substrate/frame/treasury/src/lib.rs:125]: > I would rather come up with new name, and not reuse the one which is used for local block number type

I am actually doing it the other way around calling the local one as `SystemBlockNumberFor`.

Comment (gui1117) [reply substrate/frame/treasury/src/lib.rs:125]: yes but this forces you to not import `use frame_system::pallet_prelude::*;`

I think having a different name that doesn't conflict with `frame_system::BlockNumberFor` would be a bit cleaner.

maybe like `ProvidedBlockNumberFor`? I agree the name is not good either.

anyway everything is good to me.

Comment (gupnik) [reply substrate/frame/treasury/src/lib.rs:125]: > yes but this forces you to not import `use frame_system::pallet_prelude::*;`

Yeah, my hope is to be able to provide a different type in the prelude once enough pallets have been migrated.

**Pull request merged**